Added component camect, which streams camera video using WebRTC

camect opened the pull request:
## Description:
Integrate with Camect, an affordable smart NVR, which cares about your privacy. See https://camect.com/ for more information.
This PR adds a new platform to the camera component so users could view cameras supported by Camect. In the future, we'll provide instructions to add a custom card so users will be able to view live video / footage / alerts using WebRTC.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
camect:
    host: camect.local
    port: 8443
    username: YOUR_USERNAME
    password: YOUR_PASSWORD
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

chao-camect committed:
Added component camect, which supports streaming camera video using WebRTC.

homeassistant commented:
Hi @camect, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#22446).

Once you do that we will be able to review and accept this pull request.

Thanks!

chao-camect committed:
Fix some style problems reported by the bot.

chao-camect committed:
Fix yet another style issue.

klaasnicolaas commented:
https://github.com/home-assistant/home-assistant.io/pull/9049

awarecan commented:
Thanks for willing to integrate your product into Home Assistant.

However, your integration doesn't implement the most important HA camera's method `camera_image()`, so our user cannot see your camera through HA, but only through your own interface. If my understanding is right, why you want to integrate with HA at first place?

camect commented:
@awarecan Thanks for the quick response. We don't do the usual image camera way.
We use WebRTC to stream video from Camect Home device. The code in this pull request is mainly to help HA UI connect to Camect Home.
There is another change to project home-assistant-polymer to create a card for Camect camera. I am having a difficult time sync the code to HEAD now. Hope it'll resolve soon.

awarecan commented:
This is not how Home Assistant works. We don't create a front end card only for one device. 

I am not sure if you have look at our new `stream` component debuted in release 0.90. WebRTC definitely is something we would like to bring in HA, but I would like to see a similar implement likes `stream` component for `WebRTC` component, e.g., we should be able to have a WebRTC support in our frontend, allow user to have real time communication with other family member etc. For the MVP(minimum viable product), I would like to see user can see any camera video source in a WebRTC client living in our frontend. 

I would not accept a WebRTC feature which only support one specific vendor's device. Of course, that is only my personal opinion, you can still seek other members support.

We had some discussion about how far we want HA goes towards a surveillance system (maybe it is your product's position), most of us feel that HA is not and should not become a surveillance system. Base on this, I would recommend you rethink how you want to interact with HA, I suggest you can develop a hass.io add-ons, to loosely integrate with HA. I strongly suggest you checkout this [MotionEye Addons](https://github.com/hassio-addons/addon-motioneye)

If you still feel that integrate into HA is important, then I would suggest you look at [ZoneMinder](https://www.home-assistant.io/components/zoneminder/) as a reference.

camect commented:
@awarecan Thanks for all these suggestions. I think we still want an integration instead of an addon.
Camect Home does some complicated video analysis and AI. It's not likely to work well in most user's environment. I understand why you don't want a new card. Here is my proposal:
- I am going to follow the ZoneMinder way: implement the camera_image method so users could see snapshots like other cameras using the default picture card
- We provide a custom card (not checked in to HA) that users put into their local www folder. With this custom card, users would be able to see the video through WebRTC and a lot of other features from Camera Home, like smart alerts and footage, etc.
What do you think of this?

awarecan commented:
It sounds good.

balloob commented on homeassistant/components/camect/camera.py:
```diff
@@ -0,0 +1,204 @@
+class CamectFontFileView(HomeAssistantView):
+    """Camect view to proxy font files to home."""
+
+    url = '/api/camect_proxy/font/{font_file}'
```
This does not belong in Home Assistant.

camect replied:
removed

balloob commented on homeassistant/components/camect/camera.py:
```diff
@@ -0,0 +1,204 @@
+class CamectBundleJsView(HomeAssistantView):
+    """Camect view to proxy embedded bundle JS to home."""
+
+    url = '/api/camect_proxy/bundle.js'
```
This does not belong in Home Assistant.

camect replied:
removed

balloob commented on homeassistant/components/camect/camera.py:
```diff
@@ -0,0 +1,204 @@
+class CamectWebsocketView(camera.CameraView):
+    """Camect view to proxy Websocket to home."""
+
+    url = '/api/camect_proxy/websocket/{entity_id}'
```
This does not belong in Home Assistant.

camect replied:
removed. These were used to support a new card.

balloob commented:
Custom cards living outside of HA is fine. Probably even the best approach as then people can mix and match with other Lovelace things. You can also consider using either `panel_custom` or `panel_iframe` to load resources from your box.

balloob commented:
Also, consider adding a description to your PR 🤷‍♂️

balloob commented on homeassistant/components/camect/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        hass, camera.DOMAIN, DOMAIN, conf.get(CONF_CAMERA_IDS), config)
+
+    # Register event listener.
+    def on_camect_event(evt):
```
Not used.

camect replied:
Removed. Thanks.

balloob commented on homeassistant/components/camect/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        elif evt_tp == ATTR_MODE:
+            pass
+        hass.bus.fire('camect_event', evt)
+    home.add_event_listener(lambda evt: hass.bus.fire('camect_event', evt))
```
We don't just want to connect the event busses. What are these events even to begin with?

This PR lacks some serious documentation and does too much at once.

camect replied:
OK. Removed the code. The events are camera alerts, like detected a car / person. I thought they could be useful for other components. What's required to add new events? I understand that this is premature to do it in this PR. Just want to know what we need to do if we want to bring these events back in the future.

balloob commented on homeassistant/components/camect/camera.py:
```diff
@@ -0,0 +1,204 @@
+from homeassistant.components import camect, camera
+from homeassistant.components.http import HomeAssistantView
+
+REQUIREMENTS = ['camect-py==0.1.0']
```
No reqs for platforms if components have them.

camect replied:
Done.

balloob commented on homeassistant/components/camect/camera.py:
```diff
@@ -0,0 +1,204 @@
+        return None
+
+    def camera_image(self):
+        return None
```
This property is required for camera platforms. Remove this platform if you don't intend on implementing it.

camect replied:
Implemented it.

balloob reviewed (changes requested):
There are a lot of weird things in this PR. From routing font + JS files, forwarding all events to our event bus and not following platform requiremenets.

We will not further review any part of this PR until you add a description to your PR about who you are, what you are planning to do, and how integrating with us fits in.

hunterjm commented:
I agree with everyone here.  Your `camect-py` package should utilize something like [aiortc](https://github.com/aiortc/aiortc) to establish a peer connection with your hardware and provide the requisite methods to fetch a still image so that we can follow the existing camera implementation.  Since we already have the dependencies bundled in HASS builds from 0.90 with our `stream` component, you should be able to use a combination of aiortc and it's dependency pyAV to process your feed.

On another note, I am looking at expanding the `stream` component to be able to serve WebRTC feeds directly for real time integration with the frontend and streaming to Alexa devices.  If you would like to collaborate on that, we could also expand the capabilities of `stream` to include consuming WebRTC feeds.

chao-camect committed:
Remove code supporting camect card. Will be moved to a custom card.

camect commented:
> There are a lot of weird things in this PR. From routing font + JS files, forwarding all events to our event bus and not following platform requiremenets.
Fixed all these.
> 
> We will not further review any part of this PR until you add a description to your PR about who you are, what you are planning to do, and how integrating with us fits in.
Added a description to the PR. Sorry I missed it. Let me know if more details are wanted. Thanks.

chao-camect committed:
Fix a style issue.

camect commented:
@hunterjm Thanks for the suggestions. To make sure I understand it correctly, do you plan to support WebRTC in stream component so it could feed RTSP/HLS video to Alexa using WebRTC? Or the other direction, consume WebRTC and convert that to HLS so frontend could stream it?
For the latter, why not just let frontend talk to the WebRTC provider directly?

chao-camect committed:
Merge branch 'dev' into dev

chao-camect committed:
Fix those reported by static-check and use camect-py 0.1.2.

chao-camect committed:
Add manifest.json for component camect.

chao-camect committed:
Update CODEOWNERS.

chao-camect committed:
Merge branch 'dev' into dev

chao-camect committed:
Update requirements_all.txt.

chao-camect committed:
Upgrade camect-py to 0.1.3 to remove variable type annotation, which is only supported by py3.6 and later.

camect commented:
After talked with our users, we think we'll go with a custom component instead. This way, we'll be able to provide more functionality like alerts to users. I'll close this thread for now. Thanks for all your help!
